# Exploring Ebay Car Sales Data 

This is a guided project from dataquest, designed to delvelop the skills used for exploring data using pandas and NumPY libraries, as well as data cleaning. The dataset used in this project is from eBay Kleinanzeigen, a classifieds section of the German eBay website, with 50,000 data points sampled from the original dataset to ensure fast running code, and the dataset 'dirtied' a bit, to better resemble a scraped dataset

### Displaying data

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Analysing the data:

* We can see using the DataFrame.info() method that the auto.csv file has 20 columns:

    * With 15 columns containing strings 
    
* It can also be seen that some columns have null values, but none with more than ~20% null values 
* The column names use camelcase instead of snakecase, which is preferred when using Python, meaning we can't just replace spaces with underscores

In [3]:
autos.head(1)

dateCrawled                              name  seller offerType  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat   Angebot   

    price   abtest vehicleType  yearOfRegistration  gearbox  powerPS   model  \
0  $5,000  control         bus                2004  manuell      158  andere   

    odometer  monthOfRegistration fuelType    brand notRepairedDamage  \
0  150,000km                    3      lpg  peugeot              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-26 00:00:00             0       79588  2016-04-06 06:45:54

* The DataFrame.head() method by default shows the top5 rows of the autos.csv file

In [4]:
autos.columns    

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### Renaming columns

The DataFrame.columns attribute depicts an array of the existing column names, so that we can provide the following edits to the column names using the DataFrame.rename() method

In [5]:
autos.rename(columns={
    'dateCrawled':'date_crawled',
    'offerType':'offer_type',
    'vehicleType':'vehicle_type',
    'yearOfRegistration':'registration_year',
    'powerPS':'power_ps',
    'monthOfRegistration':'registration_month',
    'fuelType':'fuel_type',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'nrOfPictures':'num_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'
}, inplace=True)

autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price   abtest vehicle_type  registration_year  gearbox  power_ps   model  \
0  $5,000  control          bus               2004  manuell       158  andere   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   

            ad_created  num_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00             0        79588  2016-04-06 06:45:54

The column names were changed from camelcase to snakecase:

* Changing all uppercase letters to lowercase 
* Using underscores to separate words

The DataFrame.describe() method with the include='all' parameter gives a good enough depiction of the data to identify areas to clean up

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Analysing the data:

* The seller and offer_type columns are identical, with the only   difference being the name
* The price and odometer columns are strings instead of integers
* The num_pictures column is filled with zeros
    
* The seller, offer_type and num_pictures columns can be dropped as they provide limited information
* The price and odometer columns need to be transformed into an integer data type
* The registration_year column has a minimum value of 1000 and a maximum value of 9999 which are both incorrect as cars have only been around in the 1900s and 2000s.
* The registration_month column has a minimum value of 0, which is incorrect as the months range from 1-12 

### Dropping columns with mostly one value

In [7]:
autos = autos.drop(["seller","offer_type","num_pictures"], axis = 1)

In [8]:
autos.shape

(50000, 17)

The DataFrame.shape attribute shows the number of columns and rows as a tuple, and as you can see, it shows that there are now 17 columns compared to the previous 20, after dropping 3 columns

### Cleaning the string numerical data
This is done by removing the non-numeric characters and then converting the string into a numeric dtype for the price and odometer columns

In [9]:
autos["price"].unique() 

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [10]:
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(int)
                 )

In [11]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [12]:
autos["odometer"] = (autos["odometer"]
                     .str.replace("km","")
                     .str.replace(",","")
                     .astype(int)
                    )

In [13]:
autos.rename(columns={"odometer":"odometer_km"}, inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Exploring price and odometer_km data

In [14]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [15]:
autos["price"].value_counts().sort_index(ascending=False).head(50)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
Name: price, dtype: int64

In [16]:
autos["price"].value_counts().sort_index().head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64

In [17]:
autos["price"].value_counts().head(50)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
1100     376
1300     371
3000     365
550      356
1800     355
5500     340
1250     335
350      335
1600     327
1999     322
400      321
6500     300
1400     292
250      291
2800     291
2300     290
1350     276
1700     268
200      266
450      265
3800     264
3200     261
2900     256
1450     252
4000     246
2999     242
2600     241
5000     239
1900     239
1650     238
Name: price, dtype: int64

We can draw the following observations using the Series.value_counts(), Series.sort_index() and Series.head() methods:

* The prices appear to increase steadily up until 350000, where beyond that the prices increase exponentially
* The most frequent price listed is 0, at 1421 counts. However, it represents only 2.84% of the total price listings, so it could be considered to be removed
* A site like ebay could list car prices at 1 to initiate bidding, therefore the price column should include car prices in the range between 1 and 350000


In [18]:
autos = autos[autos["price"].between(1,350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Understanding the date range the date columns cover

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

* The first 10 characters represent the day (2016-03-26)
* In order to understand the date range, we need to extract the date values using Series.value_counts() to generate a distribution, and the sort by index

In [20]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [21]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [22]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Analysing the data:

* The we can assume through the date last seen column the date in which the car was sold, which is why the listing was removed
* The last three dates show a disproportionate amount of cars being sold, about x6 to x8 higher

### Dealing with incorrect registration values

* As we saw earlier, the registration_year and registration_month maximum and minimum values are incorrect

In [23]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

* A car can not be registered after the listing was seen, therefore any vehicle with a registration year above 2016 is inaccurate

In [24]:
autos.loc[autos["registration_year"]<1900, "registration_year"]

10556    1800
22316    1000
24511    1111
32585    1800
49283    1001
Name: registration_year, dtype: int64

* Only 4 entries of data below 1900s, therefore it's safe to assume these entries are outliers and remove them 

In [25]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).sort_values(ascending=False)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
          ...   
1977    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1961    0.000129
1959    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1939    0.000021
1927    0.000021
1948    0.000021
1938    0.000021
1931    0.000021
1929    0.000021
1943    0.0000

* Now to look at the registration_month column

In [26]:
autos["registration_month"].describe()

count    46681.000000
mean         5.827125
std          3.670300
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [27]:
autos["registration_month"].value_counts().sort_values(ascending=False)

3     4838
6     4119
0     4026
4     3894
5     3877
7     3722
10    3495
12    3262
9     3243
11    3229
1     3105
8     3021
2     2850
Name: registration_month, dtype: int64

* The index 0 has third highest count, with 4026 counts. However, because we can't tell what month this is we should remove it 

In [28]:
autos = autos[~(autos["registration_month"] == 0)]
autos["registration_month"].value_counts().sort_values(ascending=False)

3     4838
6     4119
4     3894
5     3877
7     3722
10    3495
12    3262
9     3243
11    3229
1     3105
8     3021
2     2850
Name: registration_month, dtype: int64

### Exploring price by brand 

* To find the mean price by brand, I'll use for loops to perform aggregation

In [29]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.208229
bmw               0.112015
opel              0.103903
mercedes_benz     0.099730
audi              0.088032
ford              0.068738
renault           0.046466
peugeot           0.030008
fiat              0.025437
seat              0.018357
skoda             0.017044
nissan            0.015473
mazda             0.014934
smart             0.014441
citroen           0.013808
toyota            0.013293
hyundai           0.010245
mini              0.009354
sonstige_autos    0.009331
volvo             0.009284
mitsubishi        0.008018
honda             0.007830
kia               0.007479
alfa_romeo        0.006611
porsche           0.006400
suzuki            0.006002
chevrolet         0.005744
chrysler          0.003470
dacia             0.002766
jeep              0.002368
daihatsu          0.002344
land_rover        0.002180
subaru            0.002016
saab              0.001711
jaguar            0.001571
daewoo            0.001500
rover             0.001313
t

* There are brands that don't occupy even 1% of the data, therefore removing them wouldn't effect our data too much

In [30]:
brand_rank = autos["brand"].value_counts(normalize=True)
relevant_brands = brand_rank[brand_rank>0.01].index


mean_brand_prices = {}

for brand in relevant_brands:
    row = autos[autos["brand"]==brand]
    mean_price = row["price"].mean()
    mean_brand_prices[brand] = int(mean_price)

mean_brand_prices

{'audi': 9776,
 'bmw': 8652,
 'citroen': 3902,
 'fiat': 2950,
 'ford': 3957,
 'hyundai': 5588,
 'mazda': 4349,
 'mercedes_benz': 8882,
 'nissan': 4955,
 'opel': 3167,
 'peugeot': 3221,
 'renault': 2616,
 'seat': 4568,
 'skoda': 6569,
 'smart': 3700,
 'toyota': 5283,
 'volkswagen': 5752}

Analysing the data:

* The cheaper brands appear to be opel, ford, renault, peugeot, fiat, smart, citreon, with renault being the cheapest brand
* The more expensive brands appear to be audi, bmw and mercedes_benz, with audi being the most expensive brand
* The rest are mid-range brands, volkswagen, seat, skoda, nissan, mazda, toyota and hyundai

### Exploring mileage

* Using the pandas series and dataframe constructors, I will dispaly the mean values for the price and average mileage to evaluate the correlation between variables, if there is any
* I will use the aggregation method similar to the mean price to calculate the mean mileage 


In [31]:
mbp_series = pd.Series(mean_brand_prices)
df = pd.DataFrame(mbp_series, columns=["mean_price"])
df

mean_price
audi                 9776
bmw                  8652
citroen              3902
fiat                 2950
ford                 3957
hyundai              5588
mazda                4349
mercedes_benz        8882
nissan               4955
opel                 3167
peugeot              3221
renault              2616
seat                 4568
skoda                6569
smart                3700
toyota               5283
volkswagen           5752

In [32]:
brand_mean_mileage = {}
for brand in relevant_brands:
    rows = autos[autos["brand"]==brand]
    mean_mileage = rows["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
brand_mean_mileage 

{'audi': 128593,
 'bmw': 132489,
 'citroen': 118658,
 'fiat': 116188,
 'ford': 124009,
 'hyundai': 105514,
 'mazda': 123202,
 'mercedes_benz': 130927,
 'nissan': 117181,
 'opel': 128752,
 'peugeot': 126269,
 'renault': 127820,
 'seat': 120791,
 'skoda': 110061,
 'smart': 98855,
 'toyota': 115432,
 'volkswagen': 128183}

In [33]:
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mbp_series = pd.Series(mean_brand_prices).sort_values(ascending=False)
popularity = pd.Series(autos["brand"].value_counts(normalize=True).sort_values(ascending=False))

brand_info = pd.DataFrame(bmm_series, columns=["mean_mileage_km"])
brand_info["mean_brand_prices_$"] = mbp_series
brand_info["popularity"] = popularity

brand_info

mean_mileage_km  mean_brand_prices_$  popularity
bmw                     132489                 8652    0.112015
mercedes_benz           130927                 8882    0.099730
opel                    128752                 3167    0.103903
audi                    128593                 9776    0.088032
volkswagen              128183                 5752    0.208229
renault                 127820                 2616    0.046466
peugeot                 126269                 3221    0.030008
ford                    124009                 3957    0.068738
mazda                   123202                 4349    0.014934
seat                    120791                 4568    0.018357
citroen                 118658                 3902    0.013808
nissan                  117181                 4955    0.015473
fiat                    116188                 2950    0.025437
toyota                  115432                 5283    0.013293
skoda                   110061                 6569    0.017044
hyundai                 105514                 5588    0.010245
smart                    98855                 3700    0.014441

Analysing the data:

* There's a correlation between brand price and the subsequent mileage, the more expensive the brand, the greater mileage the car does
* opel and renault appear to be the most economical car brands
* Whereas skoda appear to be the least economical car brand 
* The popularity of a car roughly depends on the mileage 